core/policy/oracle.go

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (C) 2020 The "MysteriumNetwork/node" Authors.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package policy
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/mysteriumnetwork/node/logconfig/httptrace"
+	"github.com/mysteriumnetwork/node/market"
+	"github.com/mysteriumnetwork/node/requests"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog/log"
+)
+
+type policySubscription struct {
+	policy      market.AccessPolicy
+	eTag        string
+	subscribers []*Repository
+}
+
+// Oracle represents async policy fetcher from TrustOracle
+type Oracle struct {
+	client             *requests.HTTPClient
+	fetchURL           string
+	fetchInterval      time.Duration
+	fetchLock          sync.RWMutex
+	fetchSubscriptions []policySubscription
+
+	fetchShutdown     chan struct{}
+	fetchShutdownOnce sync.Once
+}
+
+// NewOracle create instance of policy fetcher
+func NewOracle(client *requests.HTTPClient, policyURL string, interval time.Duration) *Oracle {
+	return &Oracle{
+		client:             client,
+		fetchURL:           policyURL,
+		fetchInterval:      interval,
+		fetchSubscriptions: make([]policySubscription, 0),
+		fetchShutdown:      make(chan struct{}),
+	}
+}
+
+// Start begins fetching policies to to subscribers
+func (pr *Oracle) Start() {
+	go pr.fetchLoop()
```

Comment:
this is redundant, rename fetchLoop to Start and allow the consumer to decide if it wants to run it synchronously or asynchronously.